### PR TITLE
[JENKINS-70906] Remove prototype from core

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-70906](https://issues.jenkins.io/browse/JENKINS-70906) and https://groups.google.com/g/jenkinsci-dev/c/DR9Rr08fd0Y/m/xDWOixaXAwAJ

This is the final PR for core to finally remove all remaining remnants of Prototype.JS

### Testing done

Code inspection, manual testing, automated tests.

### Proposed changelog entries

- Remove outdated Prototype.js library.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7781"></a>

### CODE REVIEW

File: core/src/main/resources/hudson/views/BuildButtonColumn/icon.js
Comment (basil) [inline core/src/main/resources/hudson/views/BuildButtonColumn/icon.js:13]: Yeah, looks to me like the frontend equivalent of a UDP packet :smile:

File: core/src/main/resources/hudson/diagnosis/ReverseProxySetupMonitor/resources.js
Comment (timja) [reply core/src/main/resources/hudson/diagnosis/ReverseProxySetupMonitor/resources.js:21]: > so what is the need for the catch line

https://developer.mozilla.org/en-US/docs/Web/API/fetch
> A [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch) promise only rejects when a network error is encountered (which is usually when there's a permissions issue or similar). A [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch) promise does not reject on HTTP errors (404, etc.). Instead, a then() handler must check the [Response.ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) and/or [Response.status](https://developer.mozilla.org/en-US/docs/Web/API/Response/status) properties.

This admin monitor is checking for incorrect configuration / networking issues, and as such when your Jenkins / reverse proxy / whatever is not setup correctly this can (and does) cause promise rejections in this code.

----

> by invoking the callback twice if the callback throws an exception (which the old code didn't do)

It could although all the callback does it show a warning message so I think should be okay for now?

Comment (basil) [reply core/src/main/resources/hudson/diagnosis/ReverseProxySetupMonitor/resources.js:21]: Thank you, this makes sense. Since in all other places in this diff the common convention is to not catch promise rejections, the reader of this code may be naturally wondering why this does not match the general paradigm. Since the exception to the rule is justified here, I think it would be worth adding a code comment summarizing the justification for this exception which you have so clearly written above.

**Comment by timja** [comment]
ATH Passed: https://github.com/jenkinsci/acceptance-test-harness/pull/1109
PCT Failed: https://github.com/jenkinsci/bom/pull/1968

**Comment by daniel-beck** [comment]
> projectViewNested.js <-- Unsure how to test this currently

https://github.com/search?q=repo%3Ajenkinsci%2Fjenkins%20projectViewNested&type=code looks like a potential leftover from https://github.com/jenkinsci/jenkins/pull/5603. https://github.com/search?q=org%3Ajenkinsci+projectView+views+language%3AXML&type=code&l=XML indicates no callers of `projectView` that set `views`, but didn't check exhaustively.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (timja) [inline war/src/main/webapp/scripts/hudson-behavior.js:1831]: need to check if this is changing from post to get

**Comment by basil** [comment]
I added a (lightly tested) commit 457d60ad51 to fix `URICheckEncodingMonitor`. It would be great if someone could figure out how to get rid of the `Ajax.Request` usage from `select.js`. In my testing, a _lot_ of stuff in the Jenkins UI depends on this: anything that creates a list of form dropdown elements. A huge amount of other testing is blocked on that, so I think it would beneficial to take care of that sooner rather than later, even if it is only draft code.

**Comment by timja** [comment]
It should be able to be adapted by calling the callbacks in the right place I think.
Alternatively changing it to a promise approach and adapting callers. The only usage I saw (at all) that uses the callback is the credentials plugin.

I'll try look at it soon.

**Comment by timja** [comment]
https://github.com/jenkinsci/jenkins/pull/7781/commits/328e71284e2d7b4397ba1b25157614f1f167171c seems to work fine, tested with credentials anyway.

**Comment by basil** [comment]
Nice! Between this PR, all my open PRs to core, https://github.com/jenkinsci/bootstrap5-api-plugin/pull/208, https://github.com/jenkinsci/credentials-plugin/pull/443, https://github.com/jenkinsci/scm-api-plugin/pull/208, https://github.com/jenkinsci/script-security-plugin/pull/501, and https://github.com/jenkinsci/timestamper-plugin/pull/248, Jenkins is actually somewhat usable without Prototype. :sweat_smile:

**Comment by timja** [comment]
Something has broken `f:repeatable`, reproducible on configuring a new log recorder

**Comment by timja** [comment]
Fixed in https://github.com/jenkinsci/jenkins/pull/7781/commits/43ced8a34b6db5a4a164b2ba51c99b718973a107

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:233]: This is a breaking change for any current callers of this method (e.g., `matrix-auth`).

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:233]: this was adjusted in the separated PR which has been merged, I'll rebase against master and it'll be sorted

**Comment by basil** [comment]
BOM passed (pending https://github.com/jenkinsci/workflow-support-plugin/pull/221 and https://github.com/jenkinsci/matrix-project-plugin/pull/168), needs ATH

File: core/src/main/java/jenkins/model/experimentalflags/RemovePrototypeUserExperimentalFlag.java
Comment (basil) [inline core/src/main/java/jenkins/model/experimentalflags/RemovePrototypeUserExperimentalFlag.java:41]: Rather than changing the default value of the user experimental flag, I think we should completely remove Prototype, including:

- Removing the user experimental flag
- Removing `prototype.js` from the source tree
- Removing all calls to `Object.toJSON`
- Removing `Form.findMatchingInput`
- Removing the customization to `Ajax.Request.prototype.dispatchException`
- Removing `jenkins.stringify`
- Removing `ExceptionTest`

If we fail to do all of the above, we will end up in a state where we have to support both Prototype and non-Prototype in the long term. I believe it is more desirable to explicitly drop any support for Prototype to reduce long term maintenance cost.

**Review by basil — CHANGES_REQUESTED** [review]

**Review by basil — APPROVED** [review]
Also should remove the references to `prototype.js` in `.eslintignore` and `.prettierignore`.

From a technical point of view, I think this is ready to ship. Modulo https://github.com/jenkinsci/global-build-stats-plugin/pull/49 I think we are ready as we are going to be on the community side.

There are still some corporate plugins that need to be adapted (both OSS and proprietary). We may want to have a broader discussion about the timing of this change that includes these corporations. My general sense is that the corporations want clear guidance regarding when this change will be delivered so that they can plan accordingly. My general sense is also that we need to strike a balance between giving corporations enough time to adapt and delivering this change in a reasonable amount of time for the benefit of the broader Jenkins community. Whether that delivery date should be in the next month, the next three months, or beyond is a question I cannot answer with certainty at the moment.

Some corporations that I think ought to be notified:

- CloudBees
- JFrog
- Fortify/MicroFocus
- Xray/Idera
- Synopsys Coverity
- qTest/Tricentis

**Comment by MarkEWaite** [comment]
> Some corporations that I think ought to be notified:
> 
> * CloudBees
> * JFrog

I talked to @eyalbe4 and he believes that they will start the Prototype.js work in September after team members return from holidays.

CloudBees is making progress on Prototype.js use in proprietary plugins but I believe they will need until at least September to complete the work.

**Review by basil — CHANGES_REQUESTED** [review]
Is there a plan to set a date and notify people as requested in https://github.com/jenkinsci/jenkins/pull/7781#pullrequestreview-1565884856?

**Comment by timja** [comment]
How does the 3rd of October sound?

- [Fortify is aware here](https://github.com/jenkinsci/fortify-plugin/issues/64#issuecomment-1654538638)
- Mark has talked to JFrog
- CloudBees is in progress
- The others have all been notified by Jira / GitHub issues but haven't acknowledged, I'm happy to email them once we have a date agreed.

**Comment by MarkEWaite** [comment]
> How does the 3rd of October sound?

I think that is a good choice.  It gives those who are implementing in September enough time to complete their implementation, testing, and release.

**Comment by timja** [comment]
I've notified all of the requested corporations (apart from CloudBees which I assume the people on this thread can).

JFrog and MicroFocus via GitHub issue.
The rest via email.

One of the idera emails bounced, none of the others have.

**Review by basil — APPROVED** [review]
Great, I think that will give everyone plenty of time to adapt to this change, or a way to communicate with us if something comes up.

**Comment by MarkEWaite** [comment]
> I've notified all of the requested corporations (apart from CloudBees which I assume the people on this thread can).

I'll continue the reminders and status checks with CloudBees.  The status check this morning shows things are on track with the CloudBees proprietary plugins.

**Review by MarkEWaite — APPROVED** [review]
This is ready to be included in the Jenkins 2.426 weekly release scheduled for Oct 3, 2023.  The [spreadsheet](https://docs.google.com/spreadsheets/d/1dpaKALZaK0_HIGy6ony3wnegr1frTg3u1lngG4KdoC8/edit#gid=0) shows progress from widely used plugins with the following notable exceptions:

* Artifactory
* Fortify
* X-ray test management
* Coverity
* qTest

Each of the organizations that maintain those plugins have been contacted multiple times.

This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.

/label ready-for-merge

**Comment by basil** [comment]
```
 11 files changed, 7 insertions(+), 6356 deletions(-)
```

:fire:

**Pull request merged**
